// @worldsibu/convector-core-model/dist/src/convector-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var yup = require("yup");
var convector_core_errors_1 = require("@worldsibu/convector-core-errors");
var convector_core_storage_1 = require("@worldsibu/convector-core-storage");
var validate_decorator_1 = require("../src/validate.decorator");
var default_decorator_1 = require("../src/default.decorator");
var validate_decorator_2 = require("../src/validate.decorator");
var required_decorator_1 = require("../src/required.decorator");
var ConvectorModel = /** @class */ (function () {
    function ConvectorModel(content) {
        if (!content) {
            return;
        }
        if (typeof content === 'string') {
            this.id = content;
            return;
        }
        this.assign(content);
    }
    ConvectorModel.getOne = function (id, type) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var content;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        type = type || this;
                        return [4 /*yield*/, convector_core_storage_1.BaseStorage.current.get(id)];
                    case 1:
                        content = _a.sent();
                        return [2 /*return*/, new type(content)];
                }
            });
        });
    };
    ConvectorModel.query = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var type, content, _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        type = this;
                        // Stupid horrible hack to find the current implementation's parent type
                        if (args[0] && args[0].prototype.__proto__.constructor === ConvectorModel) {
                            type = args.shift();
                        }
                        return [4 /*yield*/, (_a = convector_core_storage_1.BaseStorage.current).query.apply(_a, args)];
                    case 1:
                        content = _b.sent();
                        return [2 /*return*/, Array.isArray(content) ? content.map(function (c) { return new type(c); }) : new type(content)];
                }
            });
        });
    };
    ConvectorModel.getAll = function (type) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        type = type || new this('').type;
                        return [4 /*yield*/, ConvectorModel.query(this, { selector: { type: type } })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    ConvectorModel.prototype.update = function (content) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.assign(content);
                        return [4 /*yield*/, this.save()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ConvectorModel.prototype.fetch = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var content;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, convector_core_storage_1.BaseStorage.current.get(this.id)];
                    case 1:
                        content = _a.sent();
                        if (content.type !== this.type) {
                            throw new Error("Possible ID collision, element " + this.id + " of type " + content.type + " is not " + this.type);
                        }
                        this.assign(content);
                        return [2 /*return*/];
                }
            });
        });
    };
    ConvectorModel.prototype.save = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.assign(default_decorator_1.getDefaults(this), true);
                        if (!required_decorator_1.ensureRequired(this)) {
                            throw new Error("Model " + this.type + " is not complete\n" + this.toJSON());
                        }
                        convector_core_errors_1.InvalidIdError.test(this.id);
                        return [4 /*yield*/, convector_core_storage_1.BaseStorage.current.set(this.id, this)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ConvectorModel.prototype.clone = function () {
        return Object.assign({}, this);
    };
    ConvectorModel.prototype.toJSON = function (skipEmpty) {
        var _this = this;
        if (skipEmpty === void 0) { skipEmpty = false; }
        var proto = Object.getPrototypeOf(this);
        var base = Object.keys(this).concat('id')
            .filter(function (k) { return !k.startsWith('_'); })
            .filter(function (k) { return !skipEmpty || _this[k] !== undefined || _this[k] !== null; })
            .reduce(function (result, key) {
            return (tslib_1.__assign({}, result, (_a = {}, _a[key] = _this[key], _a)));
            var _a;
        }, {});
        return Object.keys(proto)
            .reduce(function (result, key) {
            var desc = Object.getOwnPropertyDescriptor(proto, key);
            var hasGetter = desc && typeof desc.get === 'function';
            if (hasGetter) {
                result[key] = desc.get.call(_this);
            }
            if (skipEmpty && (result[key] === undefined || result[key] === null)) {
                delete result[key];
            }
            return result;
        }, base);
    };
    ConvectorModel.prototype.delete = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, convector_core_storage_1.BaseStorage.current.delete(this.id)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ConvectorModel.prototype.assign = function (content, defaults) {
        if (defaults === void 0) { defaults = false; }
        var validated = ['id', 'type'].concat(validate_decorator_2.getValidatedProperties(this));
        var filteredContent = Object.keys(content)
            .map(function (key) { return key.replace(/^_/, ''); })
            .filter(function (key) { return validated.indexOf(key) >= 0; })
            .reduce(function (result, key) {
            return (tslib_1.__assign({}, result, (_a = {}, _a[key] = content[key], _a)));
            var _a;
        }, {});
        var afterDefaults = defaults ? this.toJSON(true) : {};
        Object.assign(this, filteredContent, afterDefaults);
    };
    tslib_1.__decorate([
        required_decorator_1.Required(),
        validate_decorator_1.Validate(yup.string())
    ], ConvectorModel.prototype, "id", void 0);
    return ConvectorModel;
}());
exports.ConvectorModel = ConvectorModel;
